src/form/lean.cpp

```diff
@@ -102,37 +107,173 @@ bool LeanFormula::convert(const Formula& formula, int64_t offset,
+    std::string arg = usesParameter ? "n" : "_";
+    buf << "def " << funcName << " (" << arg << " : " << domain
+        << ") : Int := " << recursiveRHS.toString(true);
+  } else {
+    // Recursive case with base cases - use pattern matching syntax
+    buf << "def " << funcName << " : " << domain << " -> Int";
+
+    // Sort base cases by constant value for consistent output
+    std::sort(baseCases.begin(), baseCases.end(),
+              [](const auto& a, const auto& b) {
+                return a.first.children.front().value <
+                       b.first.children.front().value;
+              });
+
+    // Generate pattern matching cases for base cases
+    for (const auto& bc : baseCases) {
+      const auto& constValue = bc.first.children.front().value;
+      buf << " | " << constValue.to_string() << " => "
+          << bc.second.toString(true);
+    }
+
+    // Add the recursive case
+    if (!recursiveCase.name.empty()) {
+      // For Nat domain, use n+k pattern where k is one more than the largest base case
+      if (domain == "Nat" && !baseCases.empty()) {
+        // Find the largest base case value
+        int64_t maxBaseCase = baseCases.back().first.children.front().value.asInt();
+        int64_t patternOffset = maxBaseCase + 1;
+        
+        // Transform the RHS: replace (n-k) with (n+(patternOffset-k))
+        Expression transformedRHS = recursiveRHS;
+        transformParameterReferences(transformedRHS, patternOffset, funcName);
+        
+        buf << " | n+" << patternOffset << " => " << transformedRHS.toString(true);
+      } else {
+        buf << " | n => " << recursiveRHS.toString(true);
+      }
+    }
   }
-  bool usesParameter = mainExpr.contains(Expression::Type::PARAMETER);
-  std::string arg = usesParameter ? "n" : "_";
-  buf << "def " << funcName << " (" << arg << " : " << domain
-      << ") : Int := " << mainExpr.toString(true);
   return buf.str();
 }
 
+void LeanFormula::transformParameterReferences(Expression& expr, int64_t offset,
+                                               const std::string& funcName) const {
+  // If this is a function call to funcName, transform its argument first
+  // before recursing into children
+  if (expr.type == Expression::Type::FUNCTION && expr.name == funcName) {
+    if (expr.children.size() == 1) {
+      auto& arg = expr.children[0];
+      // Check if the argument is (Int.ofNat n) + k
+      if (arg.type == Expression::Type::SUM && arg.children.size() >= 2) {
+        // Look for Int.ofNat cast and constant
+        bool hasIntOfNat = false;
+        Number constantValue = Number::ZERO;
+        std::vector<Expression> otherTerms;
+        
+        for (const auto& term : arg.children) {
+          if (term.type == Expression::Type::FUNCTION && 
+              term.name == "Int.ofNat" &&
+              term.children.size() == 1 &&
+              term.children[0].type == Expression::Type::PARAMETER) {
```

Comment:
This complex condition is duplicated code. We should use a helper function for this. Try to extend and use 
ExpressionUtil::isSimpleFunction() or a variant for it.